Resolve module paths as relative to appRootDir - for component

PradnyaBaviskar opened the pull request:
In continuation with #73,

Made modifications to component, so as to resolve the module paths relative to appRootDir.
Added test case to verify component changes.

@bajtos - Can you please review?

PradnyaBaviskar commented:
Also added test cases to verify resolving of relative paths for models. 
(Since the pull request - https://github.com/strongloop/loopback-boot/pull/110, was one commit behind the changes made for middleware, I've merged model related test cases from PL-https://github.com/strongloop/loopback-boot/pull/110 into this PL)

@bajtos - can you review?

PradnyaBaviskar commented:
Extended this PL to prioritize coffeescripts over json for all parts, viz., 
 1. bootscripts
 2. middleware and
 3. component. 

@bajtos  - can you review?

bajtos commented on lib/compiler.js:
```diff
@@ -8,6 +8,7 @@ var debug = require('debug')('loopback:boot:compiler');
 var Module = require('module');
 var _ = require('lodash');
 
+var FILE_EXTENSION_JSON = '.json';
```
Nitpick: please add an empty line to visually delimit the variable (constant) declaration from the next function.

bajtos commented on lib/compiler.js:
```diff
@@ -538,3 +548,33 @@ function resolveRelativePaths(relativePaths, appRootDir) {
   });
 }
+
+function isPreferredExtension (filename, includeExtensions, excludeExtensions) {
```
[KISS](http://en.wikipedia.org/wiki/KISS_principle) and remove `includeExtensions` and `excludeExtensions` that are not used AFAICT.

bajtos commented on lib/compiler.js:
```diff
@@ -538,3 +548,33 @@ function resolveRelativePaths(relativePaths, appRootDir) {
+  excludeExtensions = excludeExtensions || [FILE_EXTENSION_JSON];
+  var ext = path.extname(filename);
+  return ((includeExtensions.indexOf(ext) > -1) &&
+    (excludeExtensions.indexOf(ext) === -1));
```
You are converting a fast hashmap access O(1) to a slow array scan O(n). Also considering that the two arguments are not used, I find it suboptimal to convert an already-built lookup table in `require.extensions` to an array.

```js
return (ext in includeExtensions) && !(ext in excludeExtensions);
```

bajtos commented on lib/compiler.js:
```diff
@@ -402,6 +402,14 @@ function buildMiddlewareInstructions(rootDir, config) {
+          var otherFile = tryOtherExtensions(resolved.sourceFile);
+          if (otherFile !== undefined) {
+            resolved.sourceFile = otherFile;
+          }
```
This block is repeated in two places, please extract a new method.

```js
resolved.sourceFile = fixFileExtension(resolved.sourceFile);

function fixFileFileExtension(filepath) {
  if (isPreferredExtension(filepath)) return filepath;
  var otherFile = tryOtherExtensions(filepath);
  return otherFile !== undefined ? otherFile: filepath;
}
```

It may be possible to get rid of `tryOtherExtension` completely and inline it inside the new method, I'll leave that for you to decide.

bajtos replied:
What would be even better: a function called e.g. `resolveAppScriptPath` that would call `resolveAppPath` under the hood and then `fixFileExtension` too. That should reduce even more the amount of changes needed.

Since `resolveMiddlewarePath` is calling `resolveAppPath` under the hood, you can call the new method there too and that way the callers of `resolveMiddlewarePath` don't need to know about these implementation details at all.

bajtos commented on test/compiler.test.js:
```diff
@@ -514,6 +514,37 @@ describe('compiler', function() {
 
+    it('prefers coffeescript over json in `appRootDir/bootDir`', function() {
+      // add coffee-script to require.extensions
+      require('coffee-script/register');
```
Please don't register coffee script in every test. Move these two lines to the top of the test file, where other `require` calls are made.

bajtos commented on test/compiler.test.js:
```diff
@@ -514,6 +514,37 @@ describe('compiler', function() {
+
+      var instructions = boot.compile({
+        appRootDir: appdir.PATH,
+        bootDirs: ['node_modules/custom-boot']
```
This is weird. If you want to test the module-based resolution of paths then you should use `bootDirs: ['custom-boot']` only.

I don't know if the option `bootDirs` supports paths pointing to modules. It's possible that such paths are not supported, in which case this test is not necessary.

Also please keep all `bootDirs` close test together, rather than having them intertwined with other tests (e.g. for `bootScripts`).

bajtos commented on test/compiler.test.js:
```diff
@@ -662,6 +693,72 @@ describe('compiler', function() {
+
+      var instructions = boot.compile({
+        appRootDir: appdir.PATH,
+        modelSources: ['custom-models']
```
IMO, the primary purpose of this variant is to load `node_modules/custom-models` instead of `appRootDir + '/custom-models`.

When the directory does not exist in `node_modules`, then I guess it's ok to try to resolve the path as relative to `appRootDir`, although I am also ok with the solution where non-relative paths are always resolved as a module path in one of `node_modules` folders.

What I would like you to do: Either rework this test to lookup the files in `node_modules/custom-models/car*`, or add a new test that will verify this scenario.

bajtos commented on test/compiler.test.js:
```diff
@@ -662,6 +693,72 @@ describe('compiler', function() {
+      function() {
+      appdir.createConfigFilesSync({}, {}, {
+        _meta: {
+          sources: ['custom-models']
```
Same comment as above:  Either rework this test to lookup the files in `node_modules/custom-models/car*`, or add a new test that will verify this scenario.

bajtos commented on test/compiler.test.js:
```diff
@@ -1096,6 +1193,38 @@ describe('compiler', function() {
+      expect(instructions.middleware.middleware).to.have.length(1);
+      expect(instructions.middleware.middleware[0]).have.property(
+        'sourceFile', coffee);
+    });
```
Please add another unit-test to verify that coffeescript files are loaded from `node_modules` too, i.e. create `node_modules/my-middleware/index.coffee` (+json) instead of `my-middleware.coffee` (+ json).

bajtos commented on test/compiler.test.js:
```diff
@@ -1195,6 +1324,47 @@ describe('compiler', function() {
+      appdir.writeConfigFileSync('component-config.json', {
+        'component': { }
+      });
+      var appJS = appdir.writeConfigFileSync('node_modules/component.js', '');
```
Please use `node_modules/component/index.js`, that's a more typical usage.

bajtos commented on test/compiler.test.js:
```diff
@@ -1195,6 +1324,47 @@ describe('compiler', function() {
+      expect(instructions.components).to.have.length(1);
+      expect(instructions.components[0]).have.property(
+        'sourceFile', coffee);
+    });
```
Please another test to verify that `node_modules/component/component.coffee` is correctly picked too when the component-config json contains `component` key.

bajtos commented:
@PradnyaBaviskar thank you for the update, I left a bunch of comments to address. To summarise, I'd like you to:
 - Clean up the current implementation of `tryOtherExtensions`, create a new function `resolveAppScriptPath` that can be used as a replacement for `resolveAppPath` and which will deal with the coffee script extension as needed.
 - Add more tests, especially for the cases where a value `"mymodule"` should be resolved to `{appRootDir}/node_modules/mymodule/index.js` and `{appRootDir}/node_modules/mymodule/index.coffee`.

PradnyaBaviskar commented:
@bajtos - Addressed all your comments above.

Can you please review?

bajtos commented on lib/compiler.js:
```diff
@@ -513,16 +514,8 @@ function buildComponentInstructions(rootDir, componentConfig) {
-
       return {
-        sourceFile: sourceFile,
+        sourceFile: resolveAppScriptPath(rootDir, name),
```
Before this change, a component path "mymod/file.js" was never resolved as relative, i.e. "{appRootDir}/mymod/file.js".

Is there a reasonably easy way how to preserve this behaviour? For example, can we add an `options` argument to `resolveAppScriptPath` that will be passed all the way down to `tryResolveAppPath` and which will control whether values that don't start with "./" or "../" are resolved as relative or not?

Example usage here:

```js
sourceFile: resolveAppScriptPath(rootDir, name, { strict: true })
```

Thoughts?

bajtos commented on lib/compiler.js:
```diff
@@ -538,3 +531,43 @@ function resolveRelativePaths(relativePaths, appRootDir) {
+    }
+  });
+  return (results.length > 0 ? results[0] : filepath);
+}
```
This code is very similar to what we have in [loadModuleDefinition()](https://github.com/PradnyaBaviskar/loopback-boot/blob/8e9c43c16b029ee534123bf2e0b37145c335431e/lib/compiler.js#L358-L385). Can you please refactor these two methods to share as much code as possible?

I also think the approach in `loadModuleDefinition` is better because it calls `tryReadDir` only once, while `fixFileExtension` is calling it for every file being resolved.

bajtos commented on test/compiler.test.js:
```diff
@@ -662,6 +688,75 @@ describe('compiler', function() {
+      expect(instructions.models[0].sourceFile).to.equal(appJS);
+    });
+
+    it('resolves non-relative path in `modelSources` option', function() {
```
Please rename to "resolves module relative path..." to keep this consistent with other test names.

bajtos commented on test/compiler.test.js:
```diff
@@ -662,6 +688,75 @@ describe('compiler', function() {
+      expect(instructions.models[0].sourceFile).to.equal(appJS);
+    });
+
+    it('resolves non-relative path in `sources` option in `model-config.json`',
```
Please rename to "resolves module relative path..." to keep this consistent with other test names.

bajtos commented on test/compiler.test.js:
```diff
@@ -1096,6 +1191,55 @@ describe('compiler', function() {
+        'sourceFile', coffee);
+    });
+
+    it('prefers coffeescript over json for non-relative path middleware',
```
Please rename to "... for module relative middleware path" to keep this consistent with other test names.

bajtos commented on test/compiler.test.js:
```diff
@@ -1195,6 +1339,63 @@ describe('compiler', function() {
+      appdir.writeConfigFileSync('component-config.json', {
+        'index': { }
+      });
+      var appJS = appdir.writeConfigFileSync('node_modules/index.js', '');
```
It is rather unusual to have `node_modules/index.js`, typically the script is nested in a module directory, e.g. `node_modules/mycomponent/index.js`. Please rework the test to use `mycomponent` as the key in the config file and check that is is resolved to `node_modules/mycomponent/index.js`.

Please make the same change in the coffescript test below too.

bajtos commented on test/compiler.test.js:
```diff
@@ -1195,6 +1339,63 @@ describe('compiler', function() {
+        'sourceFile', coffee);
+    });
+
+    it('prefers coffeescript over json for non-relative path component',
```
Please rename to "... for module relative component path" to keep this consistent with other test names.

bajtos commented:
Reviewed, the patch looks much better now. I'd like you to refactor and simplify the implementation even more, see my comments above.

PradnyaBaviskar commented:
@bajtos, 
Updated the code, mainly to - 
 1. Refactor ```loadModuleDefinition``` and ```fixFileExtension``` functions.
 2. And secondly, added ```resolveOptions``` parameter to ```tryResolveAppPath()``` function, so as to retain backward compatibility.
 3. Also rephrased a few test cases, as suggested above.

Can you please review?

PradnyaBaviskar commented on lib/compiler.js:
```diff
@@ -538,3 +525,61 @@ function resolveRelativePaths(relativePaths, appRootDir) {
+}
+
+function fixFileExtension (filepath, files, disallowedExtensions,
+  contentDataType) {
```
Will it be a good idea, to combine `disallowedExtensions` and `contentDataType` parameters, into a single parameter like `fixOptions`. An example for `fixOptions` would be
```
{
  'disallowedExtensions': ['.node'],
  'contentDataType': 'functions' 
}
```

bajtos replied:
> a single parameter like fixOptions

Sounds good, I think the name `options` is good enough.

Few more thoughts: `.node` should be ignored everywhere we ignore `.json`, it's just another extensions that gets special treatment from Node.js.

If find `contentDataType: 'function'` difficult to understand in the context of the calling code. Perhaps `onlyScriptsExportingFunction: true` would be a better option name?

bajtos commented on lib/compiler.js:
```diff
@@ -309,9 +310,17 @@ function resolveAppPath(rootDir, relativePath) {
+
   var fullPath = path.resolve(rootDir, relativePath);
-  if (fs.existsSync(fullPath))
+  if (fs.existsSync(fullPath) && !resolveOptions.strict)
```
I think this is not correct, as you will reject absolute paths in strict mode. Perhaps the following implementation is more correct?

```js
 if (name[0] === '/') {
  fullPath = relativePath;
} else if (name.indexOf('./') === 0 || name.indexOf('../') === 0 || !resolveOptions.strict) {
  fullPath = path.resolve(rootDir, relativePath);
}

if (fullPath && fs.existsSync(fullPath)) {
  return fullPath;
}

// resolve as a module now
```

PradnyaBaviskar commented:
@bajtos,
Done with the suggested changes! 

Please review.

bajtos commented:
@slnode test please

bajtos commented:
Landed, thank you!
